### DISCUSSION TITLE
The future of the Home Assistant Community plugin

### DISCUSSION THREAD

**Discussion by tprelog** [tprelog/iocage-homeassistant#28] [General]
**UPDATE: Aug 2022** - In light of recent changes to Home Assistant, I’ve decided to have this removed from the community plugins on TrueNAS CORE. As Home Assistant continues to grow and add new features, installation on FreeBSD no longer “just works” as it had for so many years in the past. I will leave the artifact repo on GitHub for those who wish to continue using the plugin or add the files to a regular jail. However, I can not leave this so easily accessible when Home Assistant’s own default configuration no longer works as expected.

**UPDATE: Feb 02 2021** - Around December 5 2020, plugin version 5 was released.  One of the main goals was to [KISS](https://en.wikipedia.org/wiki/KISS_principle), and I'm really happy with how it came out. Since then, I decided I'm going to continue supporting the Home Assistant plugin on CORE, even if I do switch my main server to SCALE

---

Next year, [TrueNAS Scale](https://www.truenas.com/truenas-scale/) will be released - TrueNAS Scale is based on Debian and will support Docker, which brings the future of this plugin into question.

I only run <s>FreeNAS</s> TrueNAS Core (FreeBSD based) at home. When I built my server back in 2015, the use case was very simple. I wanted a computer that could run 24/7, providing a large network file share (samba) for movies / music, and hosting a multimedia server. Without even knowing that much about the  `zfs` file system, I found it to be the easiest solution for combining multiple hard drives into one large storage pool, while providing some protection against bit rot and hard drive failure. I chose FreeNAS over Linux based NAS solutions because of it's native `zfs` support.

I’m just getting my feet wet with Docker and I know even less about Kubernetes. From a non-technical stand point, I’m not so much hung up on if the underlying OS is FreeBSD (Core) or Linux (Scale). For my NAS (home file server) point of use, both Core and Scale are identical, so for me, it comes down to adding extra services to the NAS (plugins)…

I'm finding Docker easy use and I'm now _considering_ to make the jump over to TrueNAS Scale (when it's released) in order to gain native (Docker) support. I already have a test machine spun up with TN Scale Alpha running [Home Assistant Container](https://www.home-assistant.io/docs/installation/docker/) using `docker-compose`. Even in it's alpha state, TN Scale already works well enough for my simple home use.

Obviously, if I switch to using Docker on TN Scale, I will no longer be using or maintaining any iocage plugins on TrueNAS Core.   In my opinion, this plugin little more than a glorified installation script. Besides some simple convenience, I don't see much value being lost. Honestly, I have no idea how many people are using this plugin but I don't imagine it is that many.

**Comment by gajotnt** [comment]
Im using it and love the ease of use :) 
Great work, thanks :)

**Comment by csoltenborn** [comment]
I second the previous comment - I tried the plugin (which probably wouldn't have happened otherwise) and got stuck with Home Assistant although my server requirements were pretty similar to your initial ones (and I even hadn't thought about smart home stuff before, but happened to have a couple of HomematicIP devices)...

**Comment by tprelog** [comment]
I'm actively working on a major update for this plugin now. I'm hoping it will be ready before the Home Assistant Conference, on December 13 - _No_, this update is _not related_ to the Conference in anyway. I'm just excited about it and wanted to do some work on this plugin. :smile:

It's possible, this could be the last major update I make... lots of clean up is happening now. This update will be cutting ties with FreeNAS completely. Using this plugin on FreeNAS has been deprecated (it's pretty much completely broke already)

I have still not decided if I will continue to maintain this plugin after the release of TrueNAS SCALE, so this update is about really cleaning up the code, to the best of my ability, to reflect what I've learned over this time (it's not much but definitely an improvement). There's nothing new in this version except better documentation. This should (hopefully) allow this plugin to be maintained others, _if_ I decide on switching to TrueNAS SCALE... Still, that's probably at least 6 months from now.

_If_ there is enough demand, that I continue to maintain this plugin into the future, then I will consider doing so. This next update will also position this plugin for some small, but exciting improvements, if I decide to continue to maintaining this...

**Reply by BerthaG** [reply]
Thanks for all  your great job, I think you are right, and I have almost the same experience with Freenas-truenas now, i was doing a VM to install docker and it is not the best (specially for mapping drives - snapshots) So I decided to evaluate the switch to SCALE.

**Comment by csoltenborn** [comment]
Sounds great - I hope you are going to stick to the plugin :-)

Here are two ideas I had - let me know what you think about them and whether I shall open feature requests for them:

* Support for securing the connection to HA with SSL - that would be plain awesome! Note that I have no idea whether (or to which extent) this would be possible
* Support for versioning of the HA config. The idea would basically be to use `git` to version changes to the config directory - the shell menu could then offer functionalities such as
  * show an overview of the changes within a certain time period (`git`'s history)
  * restore a certain version of the config
  * compare two versions of the config
  * export a certain change (i.e., the associated commit) such that others can use or at least investigate and learn from it

My assumption is that the latter would not be very difficult to implement - on the other hand, I'm not sure how useful it actually would be ;-)

**Comment by tprelog** [comment]
**Support for securing the connection to HA with SSL - that would be plain awesome!**

Honestly, I don't even use remote access very often. When I do, it's through [Nabu Casa.](https://www.nabucasa.com/)

Well, I guess it's called Home Assistant Cloud now.

[Home Assistant Cloud](https://www.nabucasa.com/) does cost $5 US dollars per month though. I understand, not everyone is able to afford, or willing to pay an other subscription. However if it's feasible, I would encourage you to sign up. As they say on the [Home Assistant Podcast](https://hasspodcast.io/), with a subscription you can...

>**Easily connect to Google and Amazon voice assistants for a small monthly fee that also supports the Home Assistant project. Configuration is via the User Interface so no fiddling with router settings, dynamic DNS or YAML.**

I did have SSL access manually configured at some point a few years ago, so I could use my Google Assistant. However, it is not something I feel comfortable trying to provide. I don't want to be responsible for someone gaining access to your network, or you being locked out. I hope you understand.

Again, I have almost zero knowledge in the area. However _I think the best approach_ would be to [Set Up a NGNIX Reverse Proxy with SSL Terminaltion](https://www.samueldowling.com/2020/01/18/nginx-reverse-proxy-freenas-ssl-tls/) _in a separate jail_. - Please refer to the link for an explanation

---

**Support for versioning of the HA config. The idea would basically be to use git to version changes to the config directory**

**Local**

Mentioned elsewhere, I'm going to to be _strongly suggesting_ (pushing)  the use of an external dataset for configuartion. I'm going to suggest using snapshots. You do know we have snapshots, right :wink: - It's nothing I provide. It's baked right into the zfs file system used by TrueNAS. (Oh, did I mention you can browse through your snapshots from Windows!)

**Online**

You don't need anything from me to do this. I'll add `git` and `rclone` in the plugin manifest so they are available by default. You can also use `pkg install some_package` , if you ever need something I have not included. Once Home Assistant is installed, you can follow standard guides such as [sharring your configuration on GitHub](https://community.home-assistant.io/t/sharing-your-configuration-on-github/195144) or on @DubhAd's [blog post here](https://blog.ceard.tech/2017/10/backing-up-home-assistant.html) - You want to scroll down to the **Online Bbackups** section to get stared.

---

I'm gonna re-word this question and answer it again, related to the plugin version

**Support for versioning <s>of the HA config</s> of this plugin, Mostly, the idea would basically be to use git to version changes to the <s>config</s> overlay directory**

Well thank you, I'm glad you asked :smile: 

I've started another [wiki page here](https://github.com/tprelog/iocage-homeassistant/wiki/temp_about_version)

Basically, I think it would work, to tag certain plugin dates as a point release on GitHub. Then, the point release on GitHub will only be bumped when files in the overlay should be updated, for example to fix a bug or add new functionality that does not require a full Plugin > UPDATE.

Im thinking, the tagged point release on GitHub could be used by a simple shell script and command line sensor in Home Assistant, to track if there is an update for the plugin available. In theory, a point release update to the plugin, could be triggered right from Home Assistant. At very least, we could use the console menu.

**Comment by fahadshery** [comment]
Hi,

I am as excited as you're about the TrueNas Scale for pretty much the same reasons. But, like you, I haven't decided yet if I will make the jump from freebsd to Debian based Truenas scale (yet). Will have to wait and see if it will be as stable as the current FreeBsd based TrueNAS Core. 

This is the first time I was gonna start using Home Assistant and was very pleased to see the community edition. It's my standard practice to check plugin managers GitHub page and poke around to see how active it is and if there are anything I need to be aware off before installing the plugin and I'm glad I did come to your Github page and decided to wait for your major release and once its available 👍 

Any idea when your new version will be up?

Ideally, I would want to request you to continue to maintain it since you know insides of how it works and it would be a shame to lose that if you stop maintaining it. but of course the only constant in life is change :)

**Comment by fahadshery** [comment]
> * Support for securing the connection to HA with SSL - that would be plain awesome! Note that I have no idea whether (or to which extent) this would be possible

I disagree with this because I run a reverse-proxy jail to manage all my SSL certs. its way more efficient and easier to manage. Using this, I can provide SSL certs for all the jails running on my TrueNAS Core e.g. Plex, bazzar, Sonarr, Radarr etc etc.. and soon Home Assistant once the new version is out by @tprelog ;)

**Comment by adorobis** [comment]
> * Support for securing the connection to HA with SSL - that would be plain awesome! Note that I have no idea whether (or to which extent) this would be possible

But this is already possible. There are at least 2 ways for external access (from internet):
- Deploy own SSL certificate to HA instance - when you do it only https traffic to HA is possible
- Create separate jail with e.g. nginx that will have its traffic secured by https and forward trafic to an unsecured HA instance - this option is needed when some integrations (e.g. Konnected) require http access to HA. Based on [this guide](https://help.konnected.io/support/solutions/articles/32000023964-set-up-hass-io-with-secure-remote-access-using-duckdns-and-nginx-proxy) I have deployed it with own dynamic DNS using Godaddy domain and dedicated jail for NGINX

**Comment by tprelog** [comment]
> Any idea when your new version will be up?

Yesterday :wink:

**Comment by tprelog** [comment]
>Support for securing the connection to HA with SSL -

@fahadshery @adorobis

It's my impression we have all come to the same conclusion here, and that is basically that,

It would be better to secure the connection _using a separate jail_ that can not only secure the connection to Home Assistant, but other jails as well.

Is that correct?

**Comment by adorobis** [comment]
@tprelog 
Definitely for HA, for other jails I'm not fully sure as I have directly deployed SSL certificate to Nextcloud and Plex jails and for both it works great. But maybe having it done like for HA is better?

**Comment by PackElend** [comment]
FYI:
there is a discussion in [Home Assistant VM on SCALE | TrueNAS Community](https://www.truenas.com/community/threads/home-assistant-vm-on-scale.91058/#post-630905)  what touches this topic

**Comment by tprelog** [comment]
UPDATE: Feb 02 2021 - Around December 5 2020, plugin version 5 was released. One of the main goals was to KISS, and I'm (still) really happy with how it came out. Since then, I decided I'm going to continue supporting the Home Assistant plugin on CORE, even if I do switch my main server to SCALE

This brief update has been to the top of the opening comments as well

**Comment by tprelog** [comment]
**In light of recent changes to Home Assistant, I’ve decided to have this removed from the community plugins on TrueNAS CORE.** As Home Assistant continues to grow and add new features, installation on FreeBSD no longer “just works” as it had for so many years in the past. I will leave the artifact repo on GitHub for those who wish to continue using the plugin or add the files to a regular jail. However, I can not leave this so easily accessible when Home Assistant’s own default configuration no longer works as expected.

**Comment by elyobelyob** [comment]
Just want to say thanks for all the effort you put into this over the year or so I've been engrossed with it. I recently decided to jump onto a ProxMox Mini PC solution for my HA for the future. You've been a life saver with your help and really patient too!
Many thanks!
Nick

**Reply by csoltenborn** [reply]
I can only second that (and I pretty much went the same way as you)!
